8364141: Remove LockingMode related code from x86

fbredber opened the pull request:
Since the integration of [JDK-8359437](https://bugs.openjdk.org/browse/JDK-8359437) the LockingMode flag can no longer be set by the user, instead it's declared as `const int LockingMode = LM_LIGHTWEIGHT;`. This means that we can now safely remove all `LockingMode` related code from all platforms.

This PR removes `LockingMode` related code from the **x86** platform.

When all the `LockingMode` code has been removed from all platforms, we can go on and remove it from shared (non-platform specific) files as well. And finally remove the `LockingMode` variable itself.

Passes tier1-tier5 with no added problems.

fbredber commented:
/label remove hotspot
/label add hotspot-compiler
/label add hotspot-runtime

xmas92 commented on src/hotspot/cpu/x86/interp_masm_x86.cpp:
```diff
@@ -1024,100 +1024,31 @@ void InterpreterMacroAssembler::get_method_counters(Register method,
+  const Register swap_reg = rax; // Must use rax for cmpxchg instruction
+  const Register tmp_reg = rbx;
+  const Register obj_reg = c_rarg3; // Will contain the oop
+  const Register rklass_decode_tmp = rscratch1;
```
Unused variable.

xmas92 commented on src/hotspot/cpu/x86/interp_masm_x86.cpp:
```diff
@@ -1024,100 +1024,31 @@ void InterpreterMacroAssembler::get_method_counters(Register method,
-    bind(slow_case);
+  const int obj_offset = in_bytes(BasicObjectLock::obj_offset());
+  const int lock_offset = in_bytes(BasicObjectLock::lock_offset());
+  const int mark_offset = lock_offset +
+                          BasicLock::displaced_header_offset_in_bytes();
```
Unused variable.

fbredber replied:
Fixed

xmas92 reviewed (changes requested):
Nice cleanup! Some small initial comments.

All the "displaced header" comments looks out of place. Displacing the header word on the stack (in the box) was purely a LM_LEGACY thing. Now we only displace it in the ObjectMonitor which is only handled (inflation / deflation) in the C++ runtime.

There are some more `BasicLock::displaced_header_offset_in_bytes()` asserts inside the x86 code. For callers of these methods, could be removed now or when the `BasicLock` is cleaned up.

There are some unused variables because of "displaced header" code that is kept.

`fast_lock_lightweight` and `fast_unlock_lockweight` should probably be renamed `fast_lock` and `fast_unlock` to be in sync with all the comments. (Or all the comments should be updated) (Same with C2 AD instruction)

fbredber commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -219,244 +219,12 @@ inline Assembler::AvxVectorLen C2_MacroAssembler::vector_length_encoding(int vle
 void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Register rax_reg,
                                               Register t, Register thread) {
-  assert(LockingMode == LM_LIGHTWEIGHT, "must be");
+  assert(box == rbx, "Used for displaced header location");
```
Good question! Long story short: `C2_MacroAssembler::fast_unlock` used to have this comment:

`// box: box address (displaced header location), killed. Must be EAX.
`
Looking at `cmpFastLock` in the `x86_64.ad` file I saw that the box was indeed hardwired to rax (e.g. `rax_RegP box`).

Since there was no special comment above `fast_lock_lightweight` I though that I should reuse the one from the deleted `fast_lock`. Then when looking in `cmpFastLockLightweight` I saw that box was hardwired to rbx. And ta-da, there you have the reason.

But now since you asked about it, I understand that the reason it say "Must be EAX" in `fast_lock`, was because it was used in a `cmpxchgptr`. But this was only used in the legacy locking mode, which is now deleted. So I will delete the comment and the assert().

fbredber commented:
@xmas92 
> All the "displaced header" comments looks out of place. Displacing the header word on the stack (in the box) was purely a LM_LEGACY thing. Now we only displace it in the ObjectMonitor which is only handled (inflation / deflation) in the C++ runtime.
> 
> There are some more `BasicLock::displaced_header_offset_in_bytes()` asserts inside the x86 code. For callers of these methods, could be removed now or when the `BasicLock` is cleaned up.

I basically agree. It's a bit of a mess right now and arguments that are called `disp_hdr` should probably be changed to `basic_lock`. But I'd rather do that in a separate PR and have this (and all the similar other platform PRs) only handle removing of dead code due to removal of the `LockingMode` flag. It's far easier to review a PR that has just removed code, than to a one that has also refactored lots of code.

> There are some unused variables because of "displaced header" code that is kept.

Fixed those you saw, and some more that I found.

> `fast_lock_lightweight` and `fast_unlock_lockweight` should probably be renamed `fast_lock` and `fast_unlock` to be in sync with all the comments. (Or all the comments should be updated) (Same with C2 AD instruction)

Totally agree. But that also for a separate PR after the shared (non-platform specific) files has been fixed.

fbredber commented on src/hotspot/cpu/x86/sharedRuntime_x86_64.cpp:
```diff
@@ -2194,47 +2193,7 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
     // Load the oop from the handle
     __ movptr(obj_reg, Address(oop_handle_reg, 0));
```
Fixed

dholmes-ora commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -616,8 +383,39 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
+//
+// Some commentary on balanced locking:
+//
+// fast_lock and fast_unlock are emitted only for provably balanced lock sites.
```
I assume this is also correct for `lightweight_lock` and `lightweight_unlock`?

fbredber replied:
I assume the same. But I don't want to change the comments too much, since I plan to do a clean up in which all "lightweight" prefixes will be gone and we will begin talking about it as the normal locking mode. And no, I will not just rename it `normal_lock()`.

dholmes-ora commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -616,8 +383,39 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
+// interpreter - such methods won't be compiled to use fast_lock and fast_unlock.
+// The interpreter provides two properties:
+// I1:  At return-time the interpreter automatically and quietly unlocks any
+//      objects acquired the current activation (frame).  Recall that the
```
```suggestion
//      objects acquired in the current activation (frame).  Recall that the
```

fbredber replied:
Fixed

dholmes-ora commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -616,8 +383,39 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
+//      interpreter maintains an on-stack list of locks currently held by
+//      a frame.
+// I2:  If a method attempts to unlock an object that is not held by the
+//      the frame the interpreter throws IMSX.
```
```suggestion
//      frame the interpreter throws IMSX.
```

fbredber replied:
Fixed

dholmes-ora commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -616,8 +383,39 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
+// Control returns to A() and A() unlocks O.  By I1 and I2, above, we know that O
+// is still locked by A().
+//
+// The only other source of unbalanced locking would be JNI.  The "Java Native Interface:
+// Programmer's Guide and Specification" claims that an object locked by jni_monitorenter
```
```suggestion
// The only other source of unbalanced locking would be JNI.  The "Java Native Interface
// Specification" states that an object locked by JNI's_MonitorEnter
```

dholmes-ora commented on src/hotspot/cpu/x86/interp_masm_x86.cpp:
```diff
@@ -1024,100 +1024,25 @@ void InterpreterMacroAssembler::get_method_counters(Register method,
 
-    bind(slow_case);
+  // Load object pointer into obj_reg
+  movptr(obj_reg, Address(lock_reg, BasicObjectLock::obj_offset()));
```
Do you not still need the `in_bytes()` around `obj_offset()`?

fbredber replied:
I don't think so. Or at least there are lots of examples that seems to do just fine without `in_bytes()` around `obj_offset()`, like [this line](https://github.com/openjdk/jdk/blob/2a0521863ba7d9df9b4039e61b2ce6932960cd22/src/hotspot/cpu/x86/interp_masm_x86.cpp#L1073).

dholmes-ora reviewed (approved):
Looks great! I ssuspect there may be more cleanup possible down the track but for now (with whitespace disabled) this PR clearly shows the eradication of the LockingMode.

Some minor nits in pre-existing code, and a couple of queries.

Thanks

fbredber commented on src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp:
```diff
@@ -394,21 +394,21 @@ void C2_MacroAssembler::fast_lock_lightweight(Register obj, Register box, Regist
-// should not be unlocked by "normal" java-level locking and vice-versa.  The specification
-// doesn't specify what will occur if a program engages in such mixed-mode locking, however.
+// The only other source of unbalanced locking would be JNI.  The "Java Native Interface
+// Specification" states that an object locked by JNI's_MonitorEnter should not be
```
Fixed

xmas92 reviewed (approved):
Looks good. 

As long as we take a final pass after all code changes has been performed and clean up the comments, variable-, parameter- and function-names. Would be nice to end with a consistent nomenclature, and remove all outdated terms, at least w.r.t. legacy locking.

fbredber commented:
@xmas92 
> As long as we take a final pass after all code changes has been performed and clean up the comments, variable-, parameter- and function-names. Would be nice to end with a consistent nomenclature, and remove all outdated terms, at least w.r.t. legacy locking.

That is my intention with this: [Cleanup after removing LockingMode related code](https://bugs.openjdk.org/browse/JDK-8365191)

fbredber commented:
Thanks everyone for the reviews.

/integrate
